keps/sig-apps/20190410-configmap-secret-orchestration.md

```diff
@@ -0,0 +1,214 @@
+It should be possible to collect this information before implementation begins in order to avoid requiring implementors to split their attention between writing release notes and implementing the feature itself.
+KEP editors, SIG Docs, and SIG PM should help to ensure that the tone and content of the `Summary` section is useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Kubernetes added support for higher level orchestration in the form of ReplicaSets/Deployments/DaemonSets/Statefulsets. They implement best practices for things including performing rolling upgrades and ensuring the containers are immutable. The orchestration resources currently do not fully implement a flow that provides immutable infrastructure without extra careful work by the user. Config/Secrets are injected into the immutable containers but aren't tracked along with the roll-forward/roll-back. Since config/secrets can change over time, its often the case where the user updates the config/secret for the new revision of the Deployment but rollbacks fail as they now reference the new config file instead of the one they were deployed with. The other problem with config/secret orchestration is that of orchestrated updates of the immutable infra can only be triggered manually by changes. A config/secret only change can not trigger an upgrade either manually or automatically without making an unnecessarily change to the corresponding orchestration object.
+
+We would like to solve these issues by adding support to the orchestration resources to support orchestration support around config/secrets.
+
+
+## Motivation
+
+Currently there are a mishmash of workarounds to the orchestration objects not supporting orchestration of configmaps/secrets.
+
+Several different solutions or workarounds out of tree are being used. See Alternatives section for details. None of these are widely deployed. Since most workloads need this kind of orchestration, having it as a base Kubernetes functionality would be a significant benefit to users.
+
+### Goals
+
+* Enhance the deployment/statefulset/daemonset with a flag asking for the workload to be automatically updated when a specified configmap/secret changes.
+* Enhance the deployment/statefulset/daemonset with a flag asking for the specified configmap/secret's current state to be snapshotted and follow the orchestrated workload (roll-forward/roll-back)
+* Implementations of the api changes
+
+### Non-Goals
+
+* Out of tree solutions. For applications to be portable, it really needs to be in Kubernetes proper so application developers know the functionality is always there.
+
+## Proposal
+
+The ConfigMapVolumeSource and SecretVolumeSource objects will get two new, optional fields:
+ * snapshot: boolean
+ * watch: boolean
+
+Both will default to false for compatibility.
+
+snapshot can be either true or false. Watch can only be true if snapshot is also true. This ensures immutability.
```

Comment:
Can you elaborate a bit more on "snapshot" field? What scenarios would use it? Why not just use the given ConfigMap?

Reply:
you may have things that should be snapshotted, and some things that should not. For example, say you have two secrets. One contains immutable, sensitive configuration stored in a secret (say a mysql connection string with a password). This should be marked snapshot and should not be dynamically updated out to the pod. There may be another secret though that must be updateable (for example, a cert-manager managed certificate) that you always want the newest version as the older version is no longer valid. Having a flag at the volume level allows choosing.

The only determination between choosing a configmap and a secret is how sensitive the material is.

Reply:
Why does `watch` require `snapshot` to be true? Also, wouldn't it make more sense to opt into this on workload-level (Deployment, DS, Statefulset) because if immutability is wanted, we already have the immutable configmap/secret fature?

Reply:
> I'm not super excited about this, I left some comments. This needs a better API design and a GC plan to have a chance.
> 
> I'm not excited by the temporal dependencies this introduces -- the system will do different things depending on the update order between a deployment and a configmap. If you update both in a short amount of time it's particularly unclear what would happen.
> 
> I feel like the controller changes to implement this would be fairly complex. I recommend dropping the "watch" flag completely; users can touch the object with a no-op PATCH for a rollout on-demand, add that feature in a separate step.

It really was a first stab at a kep. Open to changes.

Maybe watch in addition to snapshot is biting off too much at once. It was a primary use case of the initial issue though so thought I should include it. Often KEP's don't go through unless they cover a lot of use cases it feels like. I think if I remove it someone would probably ask for it back.